### PR TITLE
[expo-linear-gradient][android] feat: Paint dithering ability to fix color banding on Android

### PR DESCRIPTION
**Pull request by alexandrius**
# Why
Gradients on Android have color banding issue. When gradient used as a screen background it becomes very obvious and doesn't look nice. It looks worse everytime after you noticing for the first time. This PR adds option to enable paint dithering for smooth gradients.

# How
Researched `Paint` class on Android. Found and tested the dithering. 

# Test Plan

- Draw component below
```jsx
<LinearGradient
  start={{ x: 0, y: 0 }}
  end={{ x: 0, y: 1 }}
  colors={['rgb(6,25,49)', 'rgb(14,64,110)', 'rgb(152,165,193)']}
  style={{ flex: 1 }}
  ditherPaintAndroid
/>
```
- toggle between `true` and `false` for `ditherPaintAndroid` prop

Unfortunately due to compression it's not well noticeable in screenshots. Both screenshots have banding issues due to compression. Screenshots do not reflect how it will look on real hardware unfortunately. Must be tested on real device to make it obvious.

| Dither | No Dither |
| --- | ----------- |
| dither | nodither |

Demonstration:

GGuXzlIXoAA1UGp

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by alexandrius** [comment]
BTW how to avoid getting this changed?

Screenshot 2024-02-18 at 20 13 22

**Comment by BLOCKMATERIAL** [comment]
upd

**Review by ide — COMMENTED** [review]
Some feedback:

- Dithering is the part that's relevant to the developer while the "paint" part isn't relevant and is an implementation detail of the Android API. So I'd suggest calling this prop `ditherAndroid`.
- Should dithering be on by default? In the context of a linear gradient, I can't imagine banding ever being desirable. Should (disabling) dithering even be an option?

@alanjhughes Could you review this one please?

**Comment by alexandrius** [comment]
> Dithering is the part that's relevant to the developer while the "paint" part isn't relevant and is an implementation detail of the Android API. So I'd suggest calling this prop ditherAndroid.

👍

> Should dithering be on by default? In the context of a linear gradient, I can't imagine banding ever being desirable. Should (disabling) dithering even be an option?

I have 2 points why we shouldn't enable it by default for now
- Color banding isn't obvious for smaller (non background) Gradients
- According to [docs](https://developer.android.com/reference/android/graphics/Paint#setDither(boolean)) enabling dithering is slower. We don't want people to suddenly experience slowdown. SDK 52 can include this by default after people are aware of change.

**Comment by ide** [comment]
What is the actual (measured) difference in performance?

File: packages/expo-linear-gradient/src/LinearGradient.tsx [hunk 1]
```diff
@@ -58,14 +58,20 @@ export type LinearGradientProps = ViewProps & {
    * @default { x: 0.5, y: 1.0 }
    */
   end?: LinearGradientPoint | null;
+
+  /**
+   * Enables or disables paint dithering on Android. Paint dithering can fix color banding issue.
+   * @default false
+   */
+  ditherPaintAndroid?: boolean | false;
```
Comment (Simek) [inline packages/expo-linear-gradient/src/LinearGradient.tsx:66]:
```suggestion

  /**
   * Enables or disables paint dithering on Android. Paint dithering can fix color banding issue.
   * @default false
   * @platform android
   */
  ditherPaint?: boolean;
```

Typing, naming and JSDoc nit. We usually do not include the platform name in platform specific props..

**Comment by Simek** [comment]
Would be nice if you can also add an example/test case using this prop to the NCL screen for the package:
* https://github.com/expo/expo/blob/main/apps/native-component-list/src/screens/LinearGradientScreen.tsx

File: packages/expo-linear-gradient/src/LinearGradient.tsx (hunk 1 above)
Comment (alanjhughes) [inline packages/expo-linear-gradient/src/LinearGradient.tsx:66]:
```suggestion
  dither?: boolean;
```

File: packages/expo-linear-gradient/android/src/main/java/expo/modules/lineargradient/LinearGradientModule.kt [hunk 2]
```diff
@@ -29,6 +29,10 @@ class LinearGradientModule : Module() {
       Prop("borderRadii") { view, borderRadii: FloatArray? ->
         view.setBorderRadii(borderRadii ?: FloatArray(8) { 0f })
       }
+
+      Prop("ditherPaintAndroid") { view, ditherPaint: Boolean ->
+        view.setDitherPaint(ditherPaint)
+      }
```
Comment (alanjhughes) [inline packages/expo-linear-gradient/android/src/main/java/expo/modules/lineargradient/LinearGradientModule.kt:35]:
```suggestion
      Prop("dither") { view, shouldDither: Boolean? ->
        view.setDitherPaint(shouldDither ?: false)
      }
```
Removing the prop will cause null to be passed here. I see you are always passing something on the JS side but it's still better to handle this in native.

**Review by alanjhughes — CHANGES_REQUESTED** [review]
I'd also be in favour of this being the default. This has been made default in other frameworks and is easy to opt out of. The gradient should look good using the defaults, if performance is a concern then just disable it. I don't see any strong performance warnings related to enabling it.

**Comment by alexandrius** [comment]
Addressed comments. 
- Enabled dithering by default with a flag
- Improved docs
- Changed from `ditherPaintAndroid` to `dither`

**Review by alanjhughes — APPROVED** [review]

**Comment by alanjhughes** [comment]
Thanks @alexandrius!

**Pull request merged**